README.md
[English](https://github.com/yh-luo/taiwan_zip/blob/main/README.md) | [正體中文](https://github.com/yh-luo/taiwan_zip/blob/main/README_zhTW.md)

![coverage][coverage_badge]

Map zip code to cities in Taiwan.

## Supported formats

This package only supports three-digits zip codes.

## Features

* Get all the zip codes in Taiwan
* Get all the cities in Taiwan
* Find the district of a zip code
* Find the three-digits zip code of a district
* Find the districts of a city

## Example

`example/main.dart`

[coverage_badge]: https://github.com/yh-luo/taiwan_zip/blob/main/coverage_badge.svg

CHANGELOG.md
# 0.4.0

**Changes**

- Fix the problem when zip code is not unique, `TaiwanZip.toDistrict` only returns the first found district.
    - The structure of the mapping `Map` is reversed: keys become values and values are keys now. That is, districts are used as keys and zip codes are the values.
    - When the zip code maps to multiple districts, they are concatenated with `/` in-between.

# 0.3.0

**Changes**

- 東沙群島 and 南沙群島 were mapped to 高雄市 and 釣魚臺 was mapped to 宜蘭縣. The number of cities is now 24.
- All the functions can be used without an instance, i.e., they are all static functions now.

**Enhancements**

- More informative exceptions
- Better documentation
- Updated README.md

# 0.2.0

**New**

* Getter `zipCodes` returns all the zip codes.
* Getter `cities` returns all the cities.

**Changes**

- `getDistricts` now returns district names without city. Previous behavior equivalent is setting `withCity = true`.

**Enhancements**

* Removed unnecessary dependencies.

# 0.1.0

* Initial development release.

test/src/taiwan_zip_test.dart
// ignore_for_file: prefer_const_constructors
// ignore_for_file: prefer_const_declarations
import 'package:taiwan_zip/taiwan_zip.dart';
import 'package:test/test.dart';

void main() {
  const Map<String, String> testCases = {
    '104': '臺北市中山區',
    '303': '新竹縣湖口鄉',
    '307': '新竹縣芎林鄉',
    '422': '臺中市石岡區',
    '552': '南投縣集集鎮',
    '646': '雲林縣古坑鄉',
    '701': '臺南市東區',
    '831': '高雄市大寮區',
  };

  const Map<String, String> specialTestCases = {
    '300': '新竹市北區/新竹市東區/新竹市香山區',
    '600': '嘉義市東區/嘉義市西區',
  };

  test('mapping is correct', () {
    expect(mapping.length, 371);
  });

  test('get zipCodes', () {
    expect(TaiwanZip.zipCodes, mapping.values.toSet().toList());
  });

  test('get cities', () {
    expect(TaiwanZip.cities.length, 22);
  });

  group('TaiwanZip', () {
    group('toDistrict', () {
      test('works properly', () {
        testCases.forEach((key, value) {
          expect(TaiwanZip.toDistrict(key), value);
        });
        specialTestCases.forEach((key, value) {
          expect(TaiwanZip.toDistrict(key), value);
        });
      });

      test('throws when zip code is invalid', () {
        expect(() => TaiwanZip.toDistrict('123'),
            throwsA(TypeMatcher<DistrictNotFoundFailure>()));
      });
    });

    group('toZip', () {
      test('works properly', () {
        testCases.forEach((key, value) {
          expect(TaiwanZip.toZip(value), key);
        });
        specialTestCases.forEach((key, value) {
          final testDistricts = value.split('/');
          for (var district in testDistricts) {
            expect(TaiwanZip.toZip(district), key);
          }
        });
      });

      test('throws when 台 is used', () {
        final district = '台北市中正區';
        expect(() => TaiwanZip.toZip(district),
            throwsA(TypeMatcher<IncorrectFormatFailure>()));
      });

      test('throws when district is invalid', () {
        final district = '測試縣無名鎮';
        expect(() => TaiwanZip.toZip(district),
            throwsA(TypeMatcher<ZipCodeNotFoundFailure>()));
      });

      test('throws when the name is not complete', () {
        final district = '中正區';
        expect(() => TaiwanZip.toZip(district),
            throwsA(TypeMatcher<IncorrectFormatFailure>()));
      });
    });

    group('getDistricts', () {
      test('works properly', () {
        final city = '臺北市';
        final actual = TaiwanZip.getDistricts(city);
        expect(actual.length, 12);
      });

      test('works properly with city', () {
        final city = '臺北市';
        final actual = TaiwanZip.getDistricts(city, withCity: true);
        expect(actual.length, 12);
        for (var district in actual) {
          expect(district.contains(city), true);
        }
      });

      test('throws when the city is invalid', () {
        final city = '臺北縣';
        expect(() => TaiwanZip.getDistricts(city),
            throwsA(TypeMatcher<DistrictNotFoundFailure>()));
      });

      test('throws with hint when 台 is provided', () {
        final city = '台北市';
        expect(() => TaiwanZip.getDistricts(city),
            throwsA(TypeMatcher<IncorrectFormatFailure>()));
      });
    });
  });
}


example/main.dart
import 'package:taiwan_zip/taiwan_zip.dart';

void main() {
  // `TaiwanZip.cities` returns all the cities in Taiwan.
  print('Cities in Taiwan: ${TaiwanZip.cities}');

  // `TaiwanZip.zipCodes` returns all the zip codes in Taiwan.
  print('There are ${TaiwanZip.zipCodes.length} unique zip codes');

  // Static functions can be used without an instance.
  // `TaiwanZip.toDistrict(zipCode)` finds the district based on the
  var zipCode = '100';
  print('$zipCode refers to ${TaiwanZip.toDistrict(zipCode)}');

  // For some cities, multiple districts use the same zip code. In such cases
  // the districts are separated with '/'.
  var specialZipCode = '300';
  print('$specialZipCode refers to multiple districts: '
      '${TaiwanZip.toDistrict(specialZipCode)}');

  // `TaiwanZip.toZip(district)` finds the zip code of a district.
  // Since the sole district names may duplicate, city prefix is required.
  var district = '臺北市中正區';
  print('$district uses ${TaiwanZip.toZip(district)} as its zip code');

  // `TaiwanZip.getDistricts(city)` returns all the districts in the city the user
  // provided.
  var city = '臺北市';
  print('Districts in $city include ${TaiwanZip.getDistricts(city)}');
}


lib/src/taiwan_zip.dart
part 'exceptions.dart';

/// The mapping rules updated in 2021/09
const Map<String, String> mapping = {
  // 臺北市
  '臺北市中正區': '100',
  '臺北市大同區': '103',
  '臺北市中山區': '104',
  '臺北市松山區': '105',
  '臺北市大安區': '106',
  '臺北市萬華區': '108',
  '臺北市信義區': '110',
  '臺北市士林區': '111',
  '臺北市北投區': '112',
  '臺北市內湖區': '114',
  '臺北市南港區': '115',
  '臺北市文山區': '116',
  // 基隆市
  '基隆市仁愛區': '200',
  '基隆市信義區': '201',
  '基隆市中正區': '202',
  '基隆市中山區': '203',
  '基隆市安樂區': '204',
  '基隆市暖暖區': '205',
  '基隆市七堵區': '206',
  // 連江縣
  '連江縣南竿鄉': '209',
  '連江縣北竿鄉': '210',
  '連江縣莒光鄉': '211',
  '連江縣東引鄉': '212',
  // 新北市
  '新北市萬里區': '207',
  '新北市金山區': '208',
  '新北市板橋區': '220',
  '新北市汐止區': '221',
  '新北市深坑區': '222',
  '新北市石碇區': '223',
  '新北市瑞芳區': '224',
  '新北市平溪區': '226',
  '新北市雙溪區': '227',
  '新北市貢寮區': '228',
  '新北市新店區': '231',
  '新北市坪林區': '232',
  '新北市烏來區': '233',
  '新北市永和區': '234',
  '新北市中和區': '235',
  '新北市土城區': '236',
  '新北市三峽區': '237',
  '新北市樹林區': '238',
  '新北市鶯歌區': '239',
  '新北市三重區': '241',
  '新北市新莊區': '242',
  '新北市泰山區': '243',
  '新北市林口區': '244',
  '新北市蘆洲區': '247',
  '新北市五股區': '248',
  '新北市八里區': '249',
  '新北市淡水區': '251',
  '新北市三芝區': '252',
  '新北市石門區': '253',
  // 宜蘭縣
  '宜蘭縣宜蘭市': '260',
  '宜蘭縣頭城鎮': '261',
  '宜蘭縣礁溪鄉': '262',
  '宜蘭縣壯圍鄉': '263',
  '宜蘭縣員山鄉': '264',
  '宜蘭縣羅東鎮': '265',
  '宜蘭縣三星鄉': '266',
  '宜蘭縣大同鄉': '267',
  '宜蘭縣五結鄉': '268',
  '宜蘭縣冬山鄉': '269',
  '宜蘭縣蘇澳鎮': '270',
  '宜蘭縣南澳鄉': '272',
  '宜蘭縣釣魚臺': '290',
  // 新竹市
  '新竹市北區': '300',
  '新竹市東區': '300',
  '新竹市香山區': '300',
  // 新竹縣
  '新竹縣竹北市': '302',
  '新竹縣湖口鄉': '303',
  '新竹縣新豐鄉': '304',
  '新竹縣新埔鎮': '305',
  '新竹縣關西鎮': '306',
  '新竹縣芎林鄉': '307',
  '新竹縣寶山鄉': '308',
  '新竹縣竹東鎮': '310',
  '新竹縣五峰鄉': '311',
  '新竹縣橫山鄉': '312',
  '新竹縣尖石鄉': '313',
  '新竹縣北埔鄉': '314',
  '新竹縣峨眉鄉': '315',
  // 桃園市
  '桃園市中壢區': '320',
  '桃園市平鎮區': '324',
  '桃園市龍潭區': '325',
  '桃園市楊梅區': '326',
  '桃園市新屋區': '327',
  '桃園市觀音區': '328',
  '桃園市桃園區': '330',
  '桃園市龜山區': '333',
  '桃園市八德區': '334',
  '桃園市大溪區': '335',
  '桃園市復興區': '336',
  '桃園市大園區': '337',
  '桃園市蘆竹區': '338',
  // 苗栗縣
  '苗栗縣竹南鎮': '350',
  '苗栗縣頭份市': '351',
  '苗栗縣三灣鄉': '352',
  '苗栗縣南庄鄉': '353',
  '苗栗縣獅潭鄉': '354',
  '苗栗縣後龍鎮': '356',
  '苗栗縣通霄鎮': '357',
  '苗栗縣苑裡鎮': '358',
  '苗栗縣苗栗市': '360',
  '苗栗縣造橋鄉': '361',
  '苗栗縣頭屋鄉': '362',
  '苗栗縣公館鄉': '363',
  '苗栗縣大湖鄉': '364',
  '苗栗縣泰安鄉': '365',
  '苗栗縣銅鑼鄉': '366',
  '苗栗縣三義鄉': '367',
  '苗栗縣西湖鄉': '368',
  '苗栗縣卓蘭鎮': '369',
  // 臺中市
  '臺中市中區': '400',
  '臺中市東區': '401',
  '臺中市南區': '402',
  '臺中市西區': '403',
  '臺中市北區': '404',
  '臺中市北屯區': '406',
  '臺中市西屯區': '407',
  '臺中市南屯區': '408',
  '臺中市太平區': '411',
  '臺中市大里區': '412',
  '臺中市霧峰區': '413',
  '臺中市烏日區': '414',
  '臺中市豐原區': '420',
  '臺中市后里區': '421',
  '臺中市石岡區': '422',
  '臺中市東勢區': '423',
  '臺中市和平區': '424',
  '臺中市新社區': '426',
  '臺中市潭子區': '427',
  '臺中市大雅區': '428',
  '臺中市神岡區': '429',
  '臺中市大肚區': '432',
  '臺中市沙鹿區': '433',
  '臺中市龍井區': '434',
  '臺中市梧棲區': '435',
  '臺中市清水區': '436',
  '臺中市大甲區': '437',
  '臺中市外埔區': '438',
  '臺中市大安區': '439',
  // 彰化縣
  '彰化縣彰化市': '500',
  '彰化縣芬園鄉': '502',
  '彰化縣花壇鄉': '503',
  '彰化縣秀水鄉': '504',
  '彰化縣鹿港鎮': '505',
  '彰化縣福興鄉': '506',
  '彰化縣線西鄉': '507',
  '彰化縣和美鎮': '508',
  '彰化縣伸港鄉': '509',
  '彰化縣員林市': '510',
  '彰化縣社頭鄉': '511',
  '彰化縣永靖鄉': '512',
  '彰化縣埔心鄉': '513',
  '彰化縣溪湖鎮': '514',
  '彰化縣大村鄉': '515',
  '彰化縣埔鹽鄉': '516',
  '彰化縣田中鎮': '520',
  '彰化縣北斗鎮': '521',
  '彰化縣田尾鄉': '522',
  '彰化縣埤頭鄉': '523',
  '彰化縣溪州鄉': '524',
  '彰化縣竹塘鄉': '525',
  '彰化縣二林鎮': '526',
  '彰化縣大城鄉': '527',
  '彰化縣芳苑鄉': '528',
  '彰化縣二水鄉': '530',
  // 南投縣
  '南投縣南投市': '540',
  '南投縣中寮鄉': '541',
  '南投縣草屯鎮': '542',
  '南投縣國姓鄉': '544',
  '南投縣埔里鎮': '545',
  '南投縣仁愛鄉': '546',
  '南投縣名間鄉': '551',
  '南投縣集集鎮': '552',
  '南投縣水里鄉': '553',
  '南投縣魚池鄉': '555',
  '南投縣信義鄉': '556',
  '南投縣竹山鎮': '557',
  '南投縣鹿谷鄉': '558',
  // 嘉義市
  '嘉義市東區': '600',
  '嘉義市西區': '600',
  // 嘉義縣
  '嘉義縣番路鄉': '602',
  '嘉義縣梅山鄉': '603',
  '嘉義縣竹崎鄉': '604',
  '嘉義縣阿里山鄉': '605',
  '嘉義縣中埔鄉': '606',
  '嘉義縣大埔鄉': '607',
  '嘉義縣水上鄉': '608',
  '嘉義縣鹿草鄉': '611',
  '嘉義縣太保市': '612',
  '嘉義縣朴子市': '613',
  '嘉義縣東石鄉': '614',
  '嘉義縣六腳鄉': '615',
  '嘉義縣新港鄉': '616',
  '嘉義縣民雄鄉': '621',
  '嘉義縣大林鎮': '622',
  '嘉義縣溪口鄉': '623',
  '嘉義縣義竹鄉': '624',
  '嘉義縣布袋鎮': '625',
  // 雲林縣
  '雲林縣斗南鎮': '630',
  '雲林縣大埤鄉': '631',
  '雲林縣虎尾鎮': '632',
  '雲林縣土庫鎮': '633',
  '雲林縣褒忠鄉': '634',
  '雲林縣東勢鄉': '635',
  '雲林縣臺西鄉': '636',
  '雲林縣崙背鄉': '637',
  '雲林縣麥寮鄉': '638',
  '雲林縣斗六市': '640',
  '雲林縣林內鄉': '643',
  '雲林縣古坑鄉': '646',
  '雲林縣莿桐鄉': '647',
  '雲林縣西螺鎮': '648',
  '雲林縣二崙鄉': '649',
  '雲林縣北港鎮': '651',
  '雲林縣水林鄉': '652',
  '雲林縣口湖鄉': '653',
  '雲林縣四湖鄉': '654',
  '雲林縣元長鄉': '655',
  // 臺南市
  '臺南市中西區': '700',
  '臺南市東區': '701',
  '臺南市南區': '702',
  '臺南市北區': '704',
  '臺南市安平區': '708',
  '臺南市安南區': '709',
  '臺南市永康區': '710',
  '臺南市歸仁區': '711',
  '臺南市新化區': '712',
  '臺南市左鎮區': '713',
  '臺南市玉井區': '714',
  '臺南市楠西區': '715',
  '臺南市南化區': '716',
  '臺南市仁德區': '717',
  '臺南市關廟區': '718',
  '臺南市龍崎區': '719',
  '臺南市官田區': '720',
  '臺南市麻豆區': '721',
  '臺南市佳里區': '722',
  '臺南市西港區': '723',
  '臺南市七股區': '724',
  '臺南市將軍區': '725',
  '臺南市學甲區': '726',
  '臺南市北門區': '727',
  '臺南市新營區': '730',
  '臺南市後壁區': '731',
  '臺南市白河區': '732',
  '臺南市東山區': '733',
  '臺南市六甲區': '734',
  '臺南市下營區': '735',
  '臺南市柳營區': '736',
  '臺南市鹽水區': '737',
  '臺南市善化區': '741',
  '臺南市大內區': '742',
  '臺南市山上區': '743',
  '臺南市新市區': '744',
  '臺南市安定區': '745',
  // 高雄市
  '高雄市新興區': '800',
  '高雄市前金區': '801',
  '高雄市苓雅區': '802',
  '高雄市鹽埕區': '803',
  '高雄市鼓山區': '804',
  '高雄市旗津區': '805',
  '高雄市前鎮區': '806',
  '高雄市三民區': '807',
  '高雄市楠梓區': '811',
  '高雄市小港區': '812',
  '高雄市左營區': '813',
  '高雄市仁武區': '814',
  '高雄市大社區': '815',
  '高雄市岡山區': '820',
  '高雄市路竹區': '821',
  '高雄市阿蓮區': '822',
  '高雄市田寮區': '823',
  '高雄市燕巢區': '824',
  '高雄市橋頭區': '825',
  '高雄市梓官區': '826',
  '高雄市彌陀區': '827',
  '高雄市永安區': '828',
  '高雄市湖內區': '829',
  '高雄市鳳山區': '830',
  '高雄市大寮區': '831',
  '高雄市林園區': '832',
  '高雄市鳥松區': '833',
  '高雄市大樹區': '840',
  '高雄市旗山區': '842',
  '高雄市美濃區': '843',
  '高雄市六龜區': '844',
  '高雄市內門區': '845',
  '高雄市杉林區': '846',
  '高雄市甲仙區': '847',
  '高雄市桃源區': '848',
  '高雄市那瑪夏區': '849',
  '高雄市茂林區': '851',
  '高雄市茄萣區': '852',
  '高雄市東沙群島': '817',
  '高雄市南沙群島': '819',
  // 澎湖縣
  '澎湖縣馬公市': '880',
  '澎湖縣西嶼鄉': '881',
  '澎湖縣望安鄉': '882',
  '澎湖縣七美鄉': '883',
  '澎湖縣白沙鄉': '884',
  '澎湖縣湖西鄉': '885',
  // 金門縣
  '金門縣金沙鎮': '890',
  '金門縣金湖鎮': '891',
  '金門縣金寧鄉': '892',
  '金門縣金城鎮': '893',
  '金門縣烈嶼鄉': '894',
  '金門縣烏坵鄉': '896',
  // 屏東縣
  '屏東縣屏東市': '900',
  '屏東縣三地門鄉': '901',
  '屏東縣霧臺鄉': '902',
  '屏東縣瑪家鄉': '903',
  '屏東縣九如鄉': '904',
  '屏東縣里港鄉': '905',
  '屏東縣高樹鄉': '906',
  '屏東縣鹽埔鄉': '907',
  '屏東縣長治鄉': '908',
  '屏東縣麟洛鄉': '909',
  '屏東縣竹田鄉': '911',
  '屏東縣內埔鄉': '912',
  '屏東縣萬丹鄉': '913',
  '屏東縣潮州鎮': '920',
  '屏東縣泰武鄉': '921',
  '屏東縣來義鄉': '922',
  '屏東縣萬巒鄉': '923',
  '屏東縣崁頂鄉': '924',
  '屏東縣新埤鄉': '925',
  '屏東縣南州鄉': '926',
  '屏東縣林邊鄉': '927',
  '屏東縣東港鎮': '928',
  '屏東縣琉球鄉': '929',
  '屏東縣佳冬鄉': '931',
  '屏東縣新園鄉': '932',
  '屏東縣枋寮鄉': '940',
  '屏東縣枋山鄉': '941',
  '屏東縣春日鄉': '942',
  '屏東縣獅子鄉': '943',
  '屏東縣車城鄉': '944',
  '屏東縣牡丹鄉': '945',
  '屏東縣恆春鎮': '946',
  '屏東縣滿州鄉': '947',
  // 臺東縣
  '臺東縣臺東市': '950',
  '臺東縣綠島鄉': '951',
  '臺東縣蘭嶼鄉': '952',
  '臺東縣延平鄉': '953',
  '臺東縣卑南鄉': '954',
  '臺東縣鹿野鄉': '955',
  '臺東縣關山鎮': '956',
  '臺東縣海端鄉': '957',
  '臺東縣池上鄉': '958',
  '臺東縣東河鄉': '959',
  '臺東縣成功鎮': '961',
  '臺東縣長濱鄉': '962',
  '臺東縣太麻里鄉': '963',
  '臺東縣金峰鄉': '964',
  '臺東縣大武鄉': '965',
  '臺東縣達仁鄉': '966',
  // 花蓮縣
  '花蓮縣花蓮市': '970',
  '花蓮縣新城鄉': '971',
  '花蓮縣秀林鄉': '972',
  '花蓮縣吉安鄉': '973',
  '花蓮縣壽豐鄉': '974',
  '花蓮縣鳳林鎮': '975',
  '花蓮縣光復鄉': '976',
  '花蓮縣豐濱鄉': '977',
  '花蓮縣瑞穗鄉': '978',
  '花蓮縣萬榮鄉': '979',
  '花蓮縣玉里鎮': '981',
  '花蓮縣卓溪鄉': '982',
  '花蓮縣富里鄉': '983',
};

/// Map three-digit zip code to districts in Taiwan.
///
/// `toDistrict` returns the district based on zip code.
/// `toZip` returns the zip code based on the district.
class TaiwanZip {
  // constructor is disabled
  TaiwanZip._(); // coverage:ignore-line

  /// Returns all the zip codes.
  static List<String> get zipCodes => mapping.values.toSet().toList();

  /// Returns all the cities.
  static List<String> get cities =>
      mapping.keys.map((value) => value.substring(0, 3)).toSet().toList();

  /// Returns the corresponding district based on zip code.
  ///
  /// For some cities,the zip code maps to multiple districts. In such cases
  /// the districts are separated with '/'. For example, when the input is 300,
  /// the output is '新竹市北區/新竹市東區/新竹市香山區'.
  static String toDistrict(String zipCode) {
    var districts = [];

    if (!mapping.containsValue(zipCode)) {
      throw DistrictNotFoundFailure(zipCode);
    }

    for (var entry in mapping.entries) {
      if (entry.value == zipCode) {
        districts.add(entry.key);
      }
    }

    return districts.join('/');
  }

  /// Returns the corresponding zip code based on the district.
  ///
  /// City prefix is required because districts are not unique.
  /// For example, use `臺北市大安區` but not `大安區`.
  static String toZip(String district) {
    _checkName(district);

    if (district.length < 5) {
      throw const IncorrectFormatFailure('Use city prefix, e.g., 臺北市中正區');
    }

    for (var key in mapping.keys) {
      if (key == district) {
        return mapping[key]!;
      }
    }

    throw ZipCodeNotFoundFailure(district);
  }

  /// Get possible districts based on a city.
  ///
  /// If `withCity` is true, districts are returned with city prefix.
  /// For example,
  /// `withCity = true` returns 臺北市大安區, `withCity = false` returns 大安區.
  static List<String> getDistricts(String city, {bool withCity = false}) {
    List<String> districts;
    _checkName(city);

    final filteredDistricts =
        mapping.keys.where((value) => value.contains(city));
    if (filteredDistricts.isEmpty) {
      throw DistrictNotFoundFailure(city);
    }

    if (withCity) {
      districts = filteredDistricts.toList();
    } else {
      districts = filteredDistricts
          .map((value) => value.substring(3, value.length))
          .toList();
    }

    return districts;
  }
}

void _checkName(String name) {
  if (name.contains('台')) {
    throw const IncorrectFormatFailure('Use 臺 instead of 台');
  }
}
